### ISSUE TITLE
Move SynthSeg-like GMM sampling to LUT

### ISSUE THREAD

**Issue by vcasellesb** [TorchIO-project/torchio#1486]
Hi! I was trying to implement in pytorch SynthSeg's approach to generating synthetic scans by sampling from a randomly-parameterized GMM conditioned on a label map, and I came across your code.

I saw that your code iterates for each class value, sampling from a random normal distribution at each iteration, and assigning the values to the area covered by that class to the result.

In contrast, SynthSeg uses a look-up table (LUT) which although it's difficult to understand (at least for me), it is way faster, specially in pytorch with advanced indexing.

Is there a reason why you chose the iterative approach that is currently used in torchio? Would you be interested in an implementation of a LUT-based approach?

I am pasting an example of what I mean (LUT-based approach), done by me with Gemini and taking SynthSeg's original work as a template.

Best,
Vicent

```python3
    @staticmethod
    def _sample_uniform(size, a = 0., b = 1.):
        """
        :param a: lower bound
        :param b: upper bound
        """
        return torch.rand(size=size) * (b - a) + a

    def get_parameters(self, label_map: torch.Tensor):
        class_vals = torch.unique(label_map.ravel(), sorted=True)
        max_label = class_vals[-1]
        nlabels = len(class_vals)
        means_lut = torch.zeros(size=(max_label+1,), device=label_map.device, dtype=torch.float32)
        stds_lut = means_lut.clone()

        means = self._sample_uniform((nlabels,), *self.means_bounds).to(device=label_map.device)
        stds = self._sample_uniform((nlabels,), *self.stds_bounds).to(device=label_map.device)

        means_lut[class_vals] = means
        stds_lut[class_vals] = stds
        return means_lut, stds_lut

    def __call__(self,
                 label_map: torch.Tensor
                 ) -> torch.Tensor:
        # coerce to int64 for indexing
        label_map = label_map.long()

        means_lut, stds_lut = self.get_parameters(label_map)
        means_map = means_lut[label_map]
        stds_map = stds_lut[label_map]

        return torch.randn_like(label_map, dtype=torch.float32) * stds_map + means_map
```

**Comment by romainVala** [comment]
Hi Vincent

Thanks for the suggestion,  I indeed noticed very long computing time (few minutes) when dealing with high resolution volume (1000^3) so it may be worth to implement !

Although you need to take into account the other use-case, which is to generate random contrast images from 4D Partial volume maps (PV). 
4D PV maps are like 4D onehot encoded labelmap, except that now the voxel values are in the [0 1] ranges. In the use case I do not see how to avoid the for loop (iterative approach)

So if we change as suggested there is a need to keep this functionality,  using for instance condition on the label dimension. 

Romain

PS  I'm taking this opportunity to promote my work on full head segmentation with synthetic training (like synthseg, but with a few number of subject and higher resolution .... in case you missed it  : https://github.com/romainVala/SIAM

**Comment by vcasellesb** [comment]
Hi Romain,

Yes, I am aware of your publication! Very cool stuff. However, unless I am mistaken, the current `labels_to_image` functionality in torchio doesn't implement anything relating to PV effects, or am I mistaken?

Anyhow, I am curious as to how you  have implemented the PV effects. Do you run an `argmax` with these [0,1] range probabilities, or are they used to weight the per-class intensities?

Best,
Vicent

**Comment by romainVala** [comment]
dammed you are right !!!
(We originally implemented the label2image transform, but I guess no-one is using Partial volume, this is may be why this feature was not preserve ... too  bad I do like the PV effect (as illustrated in my figure 1)


although the transform can take as input  a 4D label (C channel dimension)
https://github.com/TorchIO-project/torchio/blob/277211314b10d940da33896e50e568905be16824/src/torchio/transforms/intensity/labels_to_image.py#L190

but they it seems the code will not handel properly the 4D case



line 224
       ` mask = (label_data[:, 0:1] == label_val).to(dtype=result.dtype)`

in my old torchio version we have for the same part

                if is_discretized:
                    mask = label_map == label
                else:
                    mask = label_map[label]

with is_dicretized define as :
```
        # Find out if we face a partial-volume image or a label map.
        # One-hot-encoded label map is considered as a partial-volume image
        all_discrete = label_map.eq(label_map.float().round()).all()
        same_num_dims = label_map.squeeze().dim() < label_map.dim()
        is_discretized = all_discrete and same_num_dims
```

@fepegar  I did not noticed we lost the possibility to use Partial volume in label2image  ... do you remember what happen ?

**Comment by romainVala** [comment]
> Anyhow, I am curious as to how you have implemented the PV effects. Do you run an `argmax` with these [0,1] range probabilities, or are they used to weight the per-class intensities?

I used argmax, to compute binary output (like for the target label I use) 

and I used them to weight the per-class intensities 
 MR physic tels you that indeed a voxel containing two tissu (1 and 2)  with partial  f1 and f2, then the resulting intensity
Ivoxel = f1*I1 + f2*I2 
(where I1 and I2 and intesitie of a pure tisue1/2 voxel)
